### PR TITLE
Fix: #233. Add generic X-RateLimit-* and X-RateLimit-Reset too.

### LINKED ISSUE #233: [ratelimiting] X-RateLimit-Reset
The rate-limit plugin should return a third header for when the time-frame expires and the limit resets. This is especially important for people writing code to avoid going over their limits. 

Here's how GitHub defines the header:

| Header Name | Description |
| --- | --- |
| `X-RateLimit-Reset` | The time at which the current rate limit window resets in [UTC epoch seconds](http://en.wikipedia.org/wiki/Unix_time). |

Since we support multiple rate-limit timeframes we would need to make it the format chosen in https://github.com/Mashape/kong/pull/224

### PR DESCRIPTION
**Pull request by ioggstream**
### Summary

Return the lowest limit header as X-RateLimit-Remaining and
the associate X-RateLimit-Reset.

This implements this I-D  https://tools.ietf.org/html/draft-polli-ratelimit-headers-01
which has been recently discussed at [IETF106](http://bit.ly/2qWhwwD)

### Full changelog

* Add X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset using the lower values between the configured ones

### Issues resolved

Fix #233

### CODE REVIEW

File: kong/plugins/rate-limiting/handler.lua [hunk 1]
```diff
@@ -105,25 +111,61 @@ function RateLimitingHandler:access(conf)
...
+           ) then
+          lowest = k
+        end
       end
 
+      -- Set the normalized ratelimit values.
+      headers[RATELIMIT_LIMIT] = headers[RATELIMIT_LIMIT .. "-" .. lowest]
+      headers[RATELIMIT_REMAINING] = headers[RATELIMIT_REMAINING .. "-" .. lowest]
+      -- Set the expected reset time.
+      headers[RATELIMIT_RESET] = EXPIRATIONS[lowest] - (current_timestamp/1000 % EXPIRATIONS[lowest])
```
Comment (bungle) [inline kong/plugins/rate-limiting/handler.lua]: According to:
https://tools.ietf.org/html/draft-polli-ratelimit-headers-01

> If a response contains both the "Retry-After" and the "RateLimit- Reset" header fields, the value of "RateLimit-Reset" MUST be  consistent with the one of "Retry-After".

So I don't think it is a good idea to have `RateLimit-Reset` or `X-RateLimit-Reset` being something else than `Retry-After` or inconsistent with it.

But having specific:
`X-RateLimit-Reset-Minute` (etc.) could be an option. What do you think?

I know that the `polli` rfc is just a draft but that is the best we have. Everything else is proprietary, except perhaps `Retry-After` on `429`. And yes, we should drop `X-` prefix at some point (not now).

File: kong/plugins/rate-limiting/handler.lua [hunk 2]
```diff
@@ -105,25 +111,61 @@ function RateLimitingHandler:access(conf)
...
+        if (lowest == nil
+            or usage[lowest].remaining > remaining
+            or (usage[lowest].remaining == remaining and EXPIRATIONS[k] > EXPIRATIONS[lowest])
+           ) then
+          lowest = k
+        end
       end
 
+      -- Set the normalized ratelimit values.
+      headers[RATELIMIT_LIMIT] = headers[RATELIMIT_LIMIT .. "-" .. lowest]
```
Comment (bungle) [inline kong/plugins/rate-limiting/handler.lua]: Perhaps the generic headers should follow this (as they are new, without `X-` prefix):
https://tools.ietf.org/html/draft-polli-ratelimit-headers-01#section-6.2.4

Comment (ioggstream) [reply kong/plugins/rate-limiting/handler.lua]: @bungle yes, but we are allowed to remove the `X-` only after the draft becomes RFC.

Comment (bungle) [reply kong/plugins/rate-limiting/handler.lua]: @ioggstream, AFAIK this:
https://tools.ietf.org/html/rfc6648

has deprecated the use of `X-` in 2012?

File: kong/plugins/rate-limiting/handler.lua (hunk 1 above)
Comment (ioggstream) [reply kong/plugins/rate-limiting/handler.lua]: IIUC in this implementation `RateLimit-Reset` and `Retry-After` should match,
but let me know if I'm wrong.

> having specific `x-ratelimit-reset-minute` ...
Header suffixes (eg `-minute`) require maps/code to interpret suffixes (Eg.  `policies.EXPIRATIONS`),
so clients  a great fan.

Can you make an example of how you would use `X-RateLimit-Reset-Minute` & co?

Thanks for your support! :D

File: kong/plugins/rate-limiting/handler.lua (hunk 2 above)
Comment (ioggstream) [reply kong/plugins/rate-limiting/handler.lua]: Yes, but using `DraftPolliV1-RateLimit-Limit` was weird :)))

Comment (bungle) [reply kong/plugins/rate-limiting/handler.lua]: Perhaps we can take a risk and just use `RateLimit-Limit`, if the RFC changes it, we should change as well. But still keeping these old `X-` headers that we have currently. We can document that those generic headers follow the RFC and if RFC is changed, we will change too.

Comment (ioggstream) [reply kong/plugins/rate-limiting/handler.lua]: Seems good. Are you interested in contributing to the spec? In case write me at robipolli@gmail.com

**Comment by bungle** [comment]
Hi,

I just opened #5335 with the `generic X-RateLimit-* and X-RateLimit-Reset`. I don't want to take anything from you, but this has been stuck so long, that perhaps it will get easier in if I try to do it as I can always communicate about it internally easier than you. Please grab a t-shirt: https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributor-t-shirt, you deserve it.  You can refer this PR! Thank you!

**Comment by bungle** [comment]
the #5335 is now merged, thus closing this.

**Pull request closed without merging**